core/unix/signal.c

```diff
@@ -4072,11 +4072,15 @@ record_pending_signal(dcontext_t *dcontext, int sig, kernel_ucontext_t *ucxt,
          * pending or delete an old b/c we might mess up the state so we
          * just drop this one: should only happen for alarm signal
          */
-        (info->accessing_sigpending && !info->nested_pending_ok &&
+        (((info->accessing_sigpending && !info->nested_pending_ok) ||
+          /* It's risky to deliver to an exiting thread: data structs are in unstable
+           * states (i#4438).  Just drop if we can.
+           */
+          dcontext->is_exiting) &&
          /* we do want to report a crash in receive_pending_signal() */
          (can_always_delay[sig] ||
           is_sys_kill(dcontext, pc, (byte *)sc->SC_XSP, &frame->info)))) {
```

Comment:
Curious about this part of the condition on L4082.

What's the intended behaviour for sys kills? The comment above says `we do want to report a crash`. Does this mean it shouldn't be dropped? But the condition `|| is_sys_kill(dcontext, pc, ...` holds true for sys kills.

Reply:
A signal sent by SYS_kill is by definition an asynchronous signal and so is not what would be called a "crash".  Does that answer the question?

Reply:
Oh I see. I thought the comment was somehow linked to that part of the condition. Thanks.